addons/mass_mailing_sms/models/mailing_mailing.py

```diff
@@ -324,6 +326,43 @@ def convert_links(self):
         res.update(super(Mailing, self - sms_mailings).convert_links())
         return res
 
+    def get_extra_characters_counts(self):
+        """Get character counts for sms widget computation of characters and sms counts.
+
+        :return: Character counts used for links, formatted as `{linkCharsCount: int, optOutLinkCharsCount: int}`.
+        """
+        # opt-out is formatted as "\nSTOP SMS : [base_url]/sms/[mailing_id]/[trace_code]"
+        # links are formatted as "[base_url]/r/[link_tracker_code]/s/[sms_id]"
+        if self:
+            self.ensure_one()
+
+        self.check_access_rights('write')
+
+        query = '''
+            SELECT COALESCE(max(id), 0) FROM sms_sms 
+            UNION ALL
+            SELECT code_length FROM (
+                SELECT COALESCE(LENGTH(code), 3) AS code_length
+                FROM link_tracker_code 
+                ORDER BY LENGTH(code) DESC LIMIT 1) ltc
+        '''
+        if not self.id:
+            query += 'UNION ALL SELECT COALESCE(max(id), 0) FROM mailing_mailing'
+
+        self.env.cr.execute(query)
+        results = self.env.cr.fetchall()
+        sms_id = results[0][0]
+        max_sms_id_length = max(len(str(sms_id)), 5)  # Assumes a mailing won't be more than 100k sms at once
+        code_length = results[1][0]
+        mailing_id_placeholder = str(self.id or results[2][0] + 1)
```

Comment:
is it really worth it to bypass the ORM ? (it's called only once when the component is instanciated)

those 3 lines are similar from a functionnal POV, (just 2 more queries) but it make the code a lot more readable

```py
max_sms_id_length = max(len(str(self.env['sms.sms'].sudo().search([], order="id DESC", limit=1).id)), 5)
mailing_id_placeholder = self.id or self.env['mailing.mailing'].sudo().search([], order="id DESC", limit=1).id + 1
code_length = len(self.env['link.tracker.code'].sudo().search([], order="id DESC", limit=1).code)
```